File: slot-dart/lib/src/controllers/reels_controller.dart
library slot.controller.reels_controller;

import 'dart:async';
import 'dart:math';

import 'package:slot/src/events/reels_event.dart';

import 'package:slot/src/models/reels_model.dart';
import 'package:slot/src/models/icon_model.dart';

import 'package:slot/src/mvc/model.dart';
import 'package:slot/src/mvc/view.dart';
import 'package:slot/src/mvc/controller.dart';

import 'package:slot/src/views/reels_view.dart';
import 'package:slot/src/views/reel_view.dart';

class ReelsController extends Controller {

  static const START_DURATION = const Duration(milliseconds: 600);
  static const STOP_DURATION = const Duration(seconds: 5);

  Timer _startTimer;
  Timer _stopTimer;

  int _startReelIndex;
  int _stopReelIndex;

  bool _spinning;

  Random _random;

  ReelsController(Model model, View view) : super(model, view) {

    _startReelIndex = 0;
    _stopReelIndex = 0;
    _spinning = false;
    _random = new Random();

  }

  void startSpin() {
    print("controllers.ReelsController.startSpin");
    if (_startTimer) {

      _startTimer.cancel();

    }
    _startReelIndex = 0;

    _startNextReel(null);

    dispatchEvent(new ReelsEvent(ReelsEvent.STARTING, false));
  }

  void _startNextReel(Timer timer) {

    if (_startReelIndex < view.reelViews.length) {

      print("controllers.ReelsController._startNextReel " + _startReelIndex.toString());
      ReelView reelView = view.reelViews[_startReelIndex];
      reelView.spin();
      _startReelIndex++;

      if (_startTimer is Null || !_startTimer.isActive) {
        _startTimer = new Timer.periodic(START_DURATION, _startNextReel);

      }

    } else {

      _startTimer.cancel();
      _spinStarted();

    }

  }

  void _spinStarted() {

    print("controllers.ReelsController._spinStarted");

    if (_stopTimer is Null || !_stopTimer.isActive) {
      _stopTimer = new Timer(STOP_DURATION, stopSpin);
    }

    dispatchEvent(new ReelsEvent(ReelsEvent.STARTED, false));

    _spinning = true;

  }

  void stopSpin() {

    print("controllers.ReelsController.stopSpin");

    _stopTimer.cancel();
    _stopReelIndex = 0;
    _stopNextReel();

    dispatchEvent(new ReelsEvent(ReelsEvent.STOPPED, false));

  }

  void _stopNextReel() {

    if (_stopReelIndex < view.reelViews.length) {

      print("controllers.ReelsController._stopNextReel " + _stopReelIndex.toString());

      ReelView reelView = view.reelViews[_stopReelIndex];
      int stopPosition = reelView.model.position + _random.nextInt(reelView.model.iconModels.length);

      reelView.onStopped = _stopNextReel;
      reelView.stop(stopPosition);
      _stopReelIndex++;

    } else {

      _stopTimer.cancel();
      _spinFinished();

    }

  }

  void _spinFinished() {

    print("controllers.ReelsController.spinFinished");

    String spinResult = "";

    for (ReelView reelView in view.reelViews) {

      IconModel iconModel = reelView.iconView(2);
      spinResult += iconModel.id.toString();

    }

    dispatchEvent(new ReelsEvent(ReelsEvent.STOPPED, false));

    _spinning = false;

  }

  ReelsModel get model {

    return super.model as ReelModel;

  }

  ReelsView get view {

    return super.view as ReelsView;

  }


}

File: src/modules/controllers/reels_controller.js
var Controller = require('../mvc/controller');

var ReelsController = function(model, view) {
  Controller.call(this, model, view);
  this.this._startReelIndex = 0;
  this._stopReelIndex = 0;
  this._spinning = false;
  this._startTimer = new Tock({
    countdown: true,
    interval: 10,
    complete: this._startNextReel
  });
  this._stopTimer = new Tock({
    countdown: true,
    interval: 10,
    complete: this.stopSpin
  });

  this.onStarting = null;
  this.onStarted = null;
  this.onStopping = null;
  this.onStopped = null;
};

ReelsController.START_DURATION = 600;
ReelsController.STOP_DURATION = 5000;
ReelsController.nextInt = function(max) {
  return Math.floor(Math.random() * max);
};

ReelsController.prototype = Object.create(Controller.prototype);
ReelsController.prototype.constructor = Controller;

ReelsController.prototype.startSpin = function() {
  console.log("controllers.ReelsController.startSpin");
  this._startReelIndex = 0;
  this._startNextReel();
  if (this.onStarting === 'function') {
    this.onStarting();
  }
};

ReelsController.prototype._startNextReel = function() {
  this._startTimer.reset();
  if (this._startReelIndex < view.reelViews.length) {
    console.log("_startNextReel " + this._startReelIndex.toString());
    var reelView = this._view.reelViews[this._startReelIndex];
    reelView.spin();
    this._startReelIndex++;
    this._startTimer.start(ReelsController.START_DURATION);
  } else {
    this._spinStarted();
  }
};

ReelsController.prototype._spinStarted = function() {
  console.log("controllers.ReelsController._spinStarted");
  this._stopTimer.start(ReelsController.STOP_DURATION);
  this._spinning = true;
  if (this.onStarted === 'function') {
    this.onStarted();
  }
};

ReelsController.prototype.stopSpin = function() {
  console.log("controllers.ReelsController.stopSpin");
  this._stopTimer.reset();
  this._stopReelIndex = 0;
  this._stopNextReel();
  if (this.onStopping === 'function') {
    this.onStopping();
  }
};

ReelsController.prototype._stopNextReel = function() {
  if (this._stopReelIndex < view.reelViews.length) {
    console.log("controllers.ReelsController._stopNextReel " + this._stopReelIndex.toString());
    var reelView = view.reelViews[this._stopReelIndex];
    var stopPosition = reelView.model.position + ReelsController.nextInt(reelView.model.iconModels.length);
    reelView.onStopped = this._stopNextReel;
    reelView.stop(stopPosition);
    this._stopReelIndex++;
  } else {
    this._stopTimer.reset();
    _spinFinished();
  }
};

ReelsController.prototype._spinFinished = function() {
  console.log("controllers.ReelsController.spinFinished");
  var result = "";
  for (var reelView in view.reelViews) {
    var iconModel = reelView.iconView(2);
    result += iconModel.id.toString();
  }
  this._spinning = false;
  if (this.onStopped === 'function') {
    this.onStopped(result);
  }
};

ReelsController.prototype.spinning = function() {
  return this._spinning;
};
